Docs

MartinWohlan opened the pull request:
- Added testing to the docs
- Added .DS_Store to the gitignore

draabe commented on docs/source/tests/pytest.rst:
```diff
@@ -0,0 +1,124 @@
+        # Assert
+        assert np.array_equal(x_position, np.array([[1, 3], [5, 7]]))
+
+Fixtures are a quite powerful tool since they are modular and can also request other fixtures.
```
Maybe one more sentence here for how we use them: to get minimal examples of e.g. data-level objects such as XY, Events, or Code. Of course, they look different depending on the method tested

MartinWohlan replied:
Added: In a nutshell they can be understood as minimal examples of e.g. data-level objects such as XY, Events, or Code. But compared to the normal objects, they are much clearer and are still able to test the full functionality of the methods. Of course, they look different depending on the method tested.

draabe commented on docs/source/tests/pytest.rst:
```diff
@@ -0,0 +1,124 @@
+
+Marks
+------
+Marks can be used to categorize your tests. To do so you need to decorate the method with ``@pytest.mark.<mark_name>``. When executing the ``pytest -m <mark_name>`` command (:doc:`see here </tests/testing>`) only methods decorated with ``@pytest.mark.<mark_name>`` will be selected for the testing. This can be advantageous if you have tests that are slower because they are for example accessing a  ...
```
The `see here` link will break if joined into one file

MartinWohlan replied:
all :doc: links where changed to :refs:

draabe commented on docs/source/tests/pytest.rst:
```diff
@@ -0,0 +1,124 @@
+About The Pytest Framework
+============================
+The pytest framework provides a feature-rich, plugin-based ecosystem that helps to easily write small as well as readable tests and it can also scale to support complex functional testing. To make sure that you can use the full functionality of pytest this section provides you some conventions and commands that are useful. If you want to get more into the whole framework you can find further info ...
+As described in the :doc:`previous section </tests/testing>` pytest follows a strict naming convention for files (``test_*.py``) and methods (``def test_*()``).
```
The `previous section` link will break if joined into one file

MartinWohlan replied:
all :doc: links where changed to :refs:

draabe commented on docs/source/tests/pytest.rst:
```diff
@@ -0,0 +1,124 @@
+    [tool.pytest.ini_options]
+    markers = [
+        "<mark_name1>: description",
+        "<mark_name12: description"
```
```suggestion
        "<mark_name2: description"
```

MartinWohlan replied:
changed to mark name 2

draabe commented on docs/source/tests/pytest.rst:
```diff
@@ -0,0 +1,124 @@
+
+Testing workflow
+================
+    #. Before starting the coding session :doc:`run pytest </tests/testing>` in your terminal to see if everything works or you get some errors which have to be fixed.
```
The `run pytest` link will break if joined into one file

MartinWohlan replied:
all :doc: links where changed to :refs:

draabe commented on docs/source/tests/pytest.rst:
```diff
@@ -0,0 +1,124 @@
+================
+    #. Before starting the coding session :doc:`run pytest </tests/testing>` in your terminal to see if everything works or you get some errors which have to be fixed.
+    #. After or before writing a class or method write the according tests and fixtures to keep your test suite always up to date.
+    #. After finishing your coding session :doc:`run pytest </tests/testing>` again.
```
The `run pytest link will break if joined into one file

draabe reviewed (changes requested):
Hi @MartinWohlan! Thanks for this great summary on testing. I think its concise, descriptive (with good examples) and well written.

I've added very few comments, but I'd ask you to refactor the files the following:
- Remove the .gitignore files to avoid merge conflicts. There also is no separate .gitignore in subfolders
- Please include both .rst files that contain your documentation as chapters into the "Contributing.rst". I think its best if we keep all docs on how to contribute (and I'd say that includes testing) in a single document. New user's will appreciate that they have a single file that (hopefully) fits all their questions they have when they start to develop. For experienced users, however, all this is not new and so we keep our directories and toctrees clean.
- Once you're finished, you don't need any changes to the top-level file in terms of toctrees

I've commented on all sections where this refactor will break links in the text.

Thanks a lot!

MartinWohlan commented:
> Hi @MartinWohlan! Thanks for this great summary on testing. I think its concise, descriptive (with good examples) and well written.
> 
> I've added very few comments, but I'd ask you to refactor the files the following:
> 
> * Remove the .gitignore files to avoid merge conflicts. There also is no separate .gitignore in subfolders
> * Please include both .rst files that contain your documentation as chapters into the "Contributing.rst". I think its best if we keep all docs on how to contribute (and I'd say that includes testing) in a single document. New user's will appreciate that they have a single file that (hopefully) fits all their questions they have when they start to develop. For experienced users, however, all this is not new and so we keep our directories and toctrees clean.
> * Once you're finished, you don't need any changes to the top-level file in terms of toctrees
> 
> I've commented on all sections where this refactor will break links in the text.
> 
> Thanks a lot!

MartinWohlan commented:
> Hi @MartinWohlan! Thanks for this great summary on testing. I think its concise, descriptive (with good examples) and well written.
> 
> I've added very few comments, but I'd ask you to refactor the files the following:
> 
> * Remove the .gitignore files to avoid merge conflicts. There also is no separate .gitignore in subfolders

I wanted to automatically add the .DS_Store file to the .gitignore but was in the wrong directory, so it created a new one. I have deleted the unnecessary files.

MartinWohlan commented on docs/source/tests/pytest.rst:
```diff
@@ -0,0 +1,124 @@
+
+How to execute pytest
+=====================
+As part of the `continuous integration <https://www.atlassian.com/continuous-delivery/continuous-integration>`_ pipeline build into the floodlight repository all the tests are going to be executed when making the pull request. Irrespective of this, tests should be carried out internally on a regular basis.
```
deleted the link

MartinWohlan commented on docs/source/tests/pytest.rst:
```diff
@@ -0,0 +1,124 @@
+
+    $ pytest <filename>.py::<methode_name> # to run a specific test (<method_name>) within a module (<filename>)
+
+# Should I exclude this examples and just reference the part of the documentation of pytest which contains all these commands
```
Deleted the comment

MartinWohlan commented on docs/source/tests/pytest.rst:
```diff
@@ -0,0 +1,124 @@
+---------
+Most of the tests depend on some sort of input. With `fixtures <https://docs.pytest.org/en/6.2.x/fixture.html>`_ pytest provides a feature with which data, test doubles or some system state can be created. Fixtures are reusable and can be used for multiple tests. In order to create a fixture you have to build a function that returns the data or system state that is needed for your testing. To do  ...
+
+.. code-block:: python
```
Updated

draabe commented:
Hi, thanks for the changes! Looks all good to me.

On a side note: Please refrain from merging unrelated branches that are not necessarily needed. This clutters the PR and leads to confusing version trees (or even unnecessary merge conflicts). It's okay this time, I'll squash commit the PR and there is merge back to develop scheduled soon.

Thanks again for your addition to the docs!
